### ISSUE TITLE
[builtin] possibly remove 'vscode-builtin-markdown-language-features'

### ISSUE THREAD

**Issue by vince-fugnitto** [eclipse-theia/theia#6984]
**Description**

At the moment, in our example applications we have both `@theia/preview` and `vscode-builtin-markdown-language-features` which causes some duplication of functionality.

What is the best approach? Should we remove `@theia/preview` or should we go ahead and remove the builtin for the moment? 

_Example: 'Open Preview to the Side'_

![image](https://user-images.githubusercontent.com/40359487/73268041-33f18980-41a8-11ea-8354-27d746405b44.png)

**Comment by Leon0402** [comment]
What's the current status here? 

The theia plugin is more stable and generally preferred, but it doesn't support all features such as markdown extensions (LaTeX Math). 

We currently ship both in the AUR package, but pinned markdown-languages-features to 1.39.2. This will use the theia extension by default. When using a newer markdown-language-features plugin, this will be the default and preview won't work at all. 

The drawbacks are still that there is confusion, possible multiple markdown icons and such stuff. We would prefer to only ship one extension.

**Comment by vince-fugnitto** [comment]
@Leon0402 thanks for the discussion 👍 

> What's the current status here?

Can you elaborate, the status of what exactly? 

The issue was opened only to bring up the fact that in our examples (part of this repo and what we test with) we use both the preview and builtin extension. For an application such as the one you’re implementing you’d be free to choose either or (whichever best suits your needs), or even both if you really wanted to.

**Comment by Leon0402** [comment]
> What is the best approach? Should we remove `@theia/preview` or should we go ahead and remove the builtin for the moment?

I was referring to this. 


Isn't it a decision theia has to make whether it ships with own markdown support or use the vs code builtin? As I said the theia plugin is working good, but lacks some features. The vs code builtin on the other hand is quite broken in theia as far as I can tell (Preview button not working). 

So what's the goal? Is theia/preview meant to replace markdown-language-support or complement it? Is there any roadmap? 
As a package manager I can obviously just make a choice and select one, but personally I feel this is an upstream issue. 

Perhaps I missunderstood  issue, because I thought the question was rather why do we have two extensions, which do similar, but not the same things.

**Comment by vince-fugnitto** [comment]
> Isn't it a decision theia has to make whether it ships with own markdown support or use the vs code builtin?

I’m not sure what you mean my ship, Eclipse Theia is a teamwork, application developers are free to choose which extensions and plugins they want included in their product based on the framework.

- https://theia-ide.org/docs/composing_applications/

> I can obviously just make a choice and select one, but personally I feel this is an upstream issue

I’m not sure I understand what the upstream issue would be? As I mentioned you can use whichever best suits your use-case. The `@theia/preview` extension existed before vscode extension support existed, but now you can choose either or for your custom product. 

> Perhaps I missunderstood issue, because I thought the question was rather why do we have two extensions, which do similar, but not the same things.

That’s alright! The issue was solely to bring up the point that in the examples present in the repo (used for development and testing) we make use of both the extension and plugin.

In the long run the plugin is likely the recommended one, but it also lacks some features the extension has.

**Comment by Leon0402** [comment]
> I’m not sure what you mean my ship, Eclipse Theia is a teamwork, application developers are free to choose which extensions and plugins they want included in their product based on the framework.
> 
>     * https://theia-ide.org/docs/composing_applications/

Yeah I know there is no "Standard theia". Which makes it somewhat hard for packaging as it would be very confusing in a linux distro if there are 20 theia packages, which are basically all the same just with some plugins being different. 

So I try to maintain a "theia version" which is similar to VS Code (-> So brings the same functionality) and also contains all "theia standard plugins", but no other third party plugins

> I’m not sure I understand what the upstream issue would be?

Let me try to explain. Theia tries to be somewhat similar to Vs Code and replace it. So there is a user expectation that it provides all or most of its functionality. 
This is archieved by being compatible with plugins and using the "builtin-plugins": 
https://github.com/eclipse-theia/theia/blob/master/package.json#L86

But Theia also offers it's own plugins like `@theia/console` (Which I believe is actually built in into vs code and not a plugin) or there is `@theia/preview`, which is (at least partly) realized in vs code as plugins -> markdown-language-features. 

I try to understand the general strategy here. Why does theia offer an extension with markdown features, even though there already exists a builtin markdown plugin. 
Why doesn't theia offer plugins for other built in vs code plugins like `vscode-builtin-merge-conflict`? 
Does Theia provide (in the feature) extension alternatives for all vs code built in? 
Or is the markdown support, as you wrote, just historic and can therefore now be removed, because now vs code extensions are supported? 

The current issues are the following: 
- Markdown Language Features is partly broken -> Is this a theia issue?
- @theia/preview is not feature complete -> Is this a theia issue? 
- Both extensions cannot really be used together as they provide conflicting functionality. One cannot just use Markdown Language Features, because then the non markdown feature of @theia/preview are missing. So I don't really have the choice here

Some possible solutions I could imagine (not mutually exclusive): 
- Remove all markdown related features from  @theia/preview (if it's just historic and there is no actual reason why this exists)
- Put all markdown realted features from @theia/preview into another package @theia/markdown (If theia wants to provide markdown functionality as an alternative to the built in markdown language features)
- Make clear that @theia/preview is the way forward and add missing features to it
- Add some code so that both extensions can coexist -> For example when right cling a markdown file and using "open with" you can actually select the extension to use. The same could be done for the preview icon (instead of displaying two identical icons)
- (Generally fix whatever is making the current markdown language feature not working) 

I believe some of these would be out of the scope of this issue and should be their own issue. I will open some if I understood what way theia wants to go. 

I hope I could explain better what the current issue is, why I don't really have the choice currently and what theia could do to improve the situation.

**Comment by ndoschek** [comment]
Closing this.
With #17680 (available from 1.73.0) we deprecated @theia/preview instead, keeping vscode.markdown-language-features as the markdown preview implementation.
